### ISSUE TITLE
Markup for Selecting Activity

### ISSUE THREAD

**Issue by WardCunningham** [fedwiki/wiki-plugin-activity#4]
A conversation on the weekly hangout lead to this suggestion from Mike Caulfield:

Just to keep the conversation going. It's hardly the most important feature. But a small language might look like this.  

```
SITES journal.hapgood.net, ward.fed.wiki.org, wiki.caulfield.pw:3000
SITELIST http://example.com/resources/happening2sites.txt
NEIGHBORHOOD yes
NEWEST hour
OLDEST week
TWINS 2
SYNOPSIS no
MINE only
```
where 

SITES comma delimited list of sites incl ports
SITELIST additional sites, comma delimited, retrieved from a URL
NEIGHBORHOOD include neighborhood sites in results yes/no
NEWEST newest content to show (minute, hour, day, week, month, year)
OLDEST oldest content to show (minute, hour, day, week, month, year)
TWINS Show only pages with n number of twins
SYNOPSIS show synopsis of page in results yes/no. Use only when limiting other options like OLDEST.
MINE Show pages I already have a twin of (yes/no/only)

SITELIST in particular makes joining communities easy. Probably the key feature.
SYNOPSIS plus a NEWEST 'day' OLDEST 'day' setting might be an interesting reading experience.  A verbose feed of new stuff.
MINE of "only" with TWINS of "2" helps me see material of mine others have edited.
MINE of "no" allows me to see material I have not yet forked to my site.

**Comment by paul90** [comment]
#6 adds support for SINCE - replacement for `OLDEST` above.

**Comment by WardCunningham** [comment]
Very nice. Date format is excellent addition.

Override Mike's suggestion of OLDEST was an accident. I just type SINCE a lot in my day job. See [NRQL](https://docs.newrelic.com/docs/insights/new-relic-insights/using-new-relic-query-language/nrql-reference).

The RegEx for day names should be more selective about endings. I suggest insisting on letters rather than any character. `[a-z]*$` rather than `.*$`.

**Comment by WardCunningham** [comment]
I'm wondering if &lt;h3> is right for the status indicator, `Activity Since ...`? This makes for two heading in a row which causes one to think that they are unrelated, two sections, the first empty. Another choice might be another line of italic  messaging at the bottom. Thoughts?

![image](https://cloud.githubusercontent.com/assets/12127/6687756/1f791304-cc6b-11e4-8082-7014e53991de.png)

**Comment by paul90** [comment]
<h2> would be better - but would also need a change to the css in the client, so it is not too big.

**Comment by WardCunningham** [comment]
Hmm. GitHub does not properly escape the tag in your last comment. When I got the email from your post I thought you were shouting at me. (Regarding headings, I'm ok as is.)

![image](https://cloud.githubusercontent.com/assets/12127/6690207/585ffeec-cc7c-11e4-8b6e-1112a9d26674.png)

**Comment by WardCunningham** [comment]
We could probably generalize this neighborhood option to be selective about which neighbors. Argument could be a string to match or a regex pattern. Or such specialized lists might be hard to explain to both those who parameterize them and to those who can't make sense of the result.

**Comment by WardCunningham** [comment]
Three ideas came up in today's hangout. Each suggests a way to focus attention on specific kinds of pages which is to say ignore other kinds of pages. We'll have to design a markup that makes this clear. In this comment I only identify the use cases.

**Site Focus:** Sometimes when bringing a new site into a neighborhood one would like a view into the activity of just that one site. Perhaps the site could offer a page with an activity plugin configured to focus on other pages from the same site. Forking this page would then direct attention to its new home.

**Word Focus:** Sometimes when a community (such as class members) discuss many things there could be value in focusing on activity related to just one subject in this community. The activity plugin could be configured to look through sitemaps and select only those pages containing a specific search term. Imagine that these terms were used like twitter hash tags, but, of course, the use of hash would be just a familiar convention, not something wired into the search.

**Link Focus:** If focus is applied to select pages by either of these means, it might be possible to look beyond the focus by following links in the text. (We don't yet report outbound links in the sitemaps but we could.) It would be natural to look one link beyond the focus because these pages are clearly related. However, the broadening could be repeatedly applied possibly with page-rank style limitations so that more removed pages would need multiple citations to be included.

**Comment by paul90** [comment]
> We could probably generalize this neighborhood option to be selective about which neighbors.

By itself `NEIGHBORHOOD yes/no` is probably not that useful - however I see it being more useful when combined with `SITES` or `ROSTER` where it probably defaults to `no` rather than the more normal `yes`. 

For anybody looking for an index of their pages, `MINE` is probably a better option as it will include flags of twins. But an index will also need a `SORT` to put the list into alphabetic rather than date order.

**Comment by WardCunningham** [comment]
Ah, yes, `NEIGHBORHOOD` disables the default behavior we have now. I feel the need to think through the general algorithm the plugin will use before deciding how that algorithm will be parameterized.

There is one idea that could be great or could be terrible. I don't know which. The question is, should the plugin fetch sitemaps on it's own that that don't become part of the tab's existing neighborhood? If it did, then it could "look ahead" for possible collaborations without cluttering the neighborhood with unwanted twins. We could expect multiple instances of the plugin to manage their own views into the federation, possibly cooperating with others just enough to make sure sitemaps aren't fetched and stored redundantly. I'll refer to this idea as **Lookahead** behavior from now on in this discussion.

**Comment by WardCunningham** [comment]
In this comment I will present a general outline for how the activity plugin could operate. I will define some terminology which I might refine based on further discussion in this issue.

* The plugin will **Observe** the evolving lineup and neighborhood.
* The plugin will **Select** some observations of interest.
* The plugin will **Render** the selection in the context of all observations.
* The plugin will **Respond** to clicks and hovers consistent with the rendering.

We should be clear which phase or phases of this process we are configuring when we suggest additional phrases in the plugin's configuration markup. The outline suggests a number of additional consequences.

* The plugin itself will not **lookahead**, but it could respond to lookahead from other plugins (such as Roster) in the lineup.

* The plugin should not be configured with **lists of sites** but could refer to sites indirectly by observable information such as the page's site or the tab's origin.

* The plugin would need to find more detailed **linking** information in the neighborhood's sitemaps than we now have if it is to provide link-focus in it's selection.

* The sort order and inclusion of observable but not otherwise selected **twins** are decisions to be made in the rendering configuration.

* The **conversation-club** click behavior of adding many twins at once to the lineup, if possible, should be rendered visibly different than the one page at a time currently offered.

**Comment by paul90** [comment]
#8 is adding support for `NEIGHBORHOOD`, `TWINS`, `SORT` and `SEARCH`

**Comment by WardCunningham** [comment]
I'm thinking I will add a ROSTER command that looks left for all roster-source items and merges any of their getRoster() elements for which the ROSTER arg is a partial match. (I have some time to program but thought I should declare my intentions first.) Examples:

```
ROSTER all
ROSTER happening
ROSTER ward
```

The command will intersect the site names with those in the neighborhood. It implies `NEIGHBORHOOD no` except for the sites that it finds in rosters. Multiple ROSTER commands will be merged into one roster for the purpose of site selection. 

See https://github.com/fedwiki/wiki-plugin-roster/pull/3 for more about Roster Plugin's item interface.

**Comment by paul90** [comment]
These changes are published as + wiki-plugin-activity@0.2.3
